src/integration/resources/inprocess/cluster.go

```diff
@@ -0,0 +1,337 @@
+func (c *CoordinatorClusterOptions) Validate() error {
+	return c.Config.Validate()
+}
+
+// DBNodeClusterConfig contains the configuration for dbnodes in the
+// cluster. Must specify one of the options but not both.
+type DBNodeClusterConfig struct {
+	// ConfigString contains the configuration as a raw YAML string.
+	ConfigString string
+	// ConfigObject contains the configuration as an inflated object.
+	ConfigObject *dbcfg.Configuration
+}
+
+// Validate validates the DBNodeClusterConfig.
+func (d *DBNodeClusterConfig) Validate() error {
+	if d.ConfigString != "" && d.ConfigObject != nil {
+		return errors.New("must specify either ConfigString or ConfigObject, but not both")
+	}
+
+	if d.ConfigString == "" && d.ConfigObject == nil {
+		return errors.New("ConfigString and ConfigObject cannot both be empty")
+	}
+
+	return nil
+}
+
+// ToConfig generates a dbcfg.Configuration object from the DBNodeClusterConfig.
+func (d *DBNodeClusterConfig) ToConfig() (dbcfg.Configuration, error) {
+	if d.ConfigObject != nil {
+		return *d.ConfigObject, nil
+	}
+
+	var cfg dbcfg.Configuration
+	if err := yaml.Unmarshal([]byte(d.ConfigString), &cfg); err != nil {
+		return dbcfg.Configuration{}, err
+	}
+
+	return cfg, nil
+}
+
+// CoordinatorClusterConfig contains the configuration for coordinators in the
+// cluster. Must specify one of the options but not both.
+type CoordinatorClusterConfig struct {
+	// ConfigString contains the configuration as a raw YAML string.
+	ConfigString string
+	// ConfigObject contains the configuration as an inflated object.
+	ConfigObject *coordinatorcfg.Configuration
+}
+
+// Validate validates the CoordinatorClusterConfig.
+func (c *CoordinatorClusterConfig) Validate() error {
+	if c.ConfigString != "" && c.ConfigObject != nil {
+		return errors.New("must specify either ConfigString or ConfigObject, but not both")
+	}
+
+	if c.ConfigString == "" && c.ConfigObject == nil {
+		return errors.New("ConfigString and ConfigObject cannot both be empty")
+	}
+
+	return nil
+}
+
+// ToConfig generates a coordinatorcfg.Configuration object from the CoordinatorClusterConfig.
+func (c *CoordinatorClusterConfig) ToConfig() (coordinatorcfg.Configuration, error) {
+	if c.ConfigObject != nil {
+		return *c.ConfigObject, nil
+	}
+
+	var cfg coordinatorcfg.Configuration
+	if err := yaml.Unmarshal([]byte(c.ConfigString), &cfg); err != nil {
+		return coordinatorcfg.Configuration{}, err
+	}
+
+	return cfg, nil
+}
+
+// NewCluster creates a new M3 cluster based on the ClusterOptions provided.
+// Expects at least a coordinator and a dbnode config.
+func NewCluster(opts ClusterOptions) (resources.M3Resources, error) {
+	if err := opts.DBNode.Validate(); err != nil {
+		return nil, err
+	}
+
+	if err := opts.Coordinator.Validate(); err != nil {
+		return nil, err
+	}
+
+	logger, err := newLogger()
+	if err != nil {
+		return nil, err
+	}
+
+	var (
+		numNodes            = opts.DBNode.RF * opts.DBNode.NumInstances
```

Comment:
This might be a very basic questions. But I am a bit confused about the terms... Why RF*Instance is the number of nodes?

Reply:
NumInstances is specific to a single RF. Namely, it's the number of instances we'll shard data across in a single RF. So, let's say we use 3 instances as an example. That means we shard the entirety of the data across 3 instances for an RF. If we set our RF to 3 as well, then the number of total nodes running M3DB is 3 * 3 or RF * NumInstances.